### ISSUE TITLE
Command Palette: highlighted suggestions insufficient contrast

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#51582]
### Description

In the Command centre, users can 'highlight' suggestions by using the Down and Up arrow keys, or by hovering on them.

Worth noting this is not a real 'focus' style. Actually, focus is on the search input. The suggestions are highlighted and 'selected' via scripting. Nevertheless, since it is possible to select and activate the suggestions, users need to have a clear indication of what the highlighted / selected suggestion is.

The highlighted state style only relies on a subtle color change. Color only is not sufficient. Teh highlighted style needs a shape, whether it's an outline or a left border like in the classic admin menu.

Aside: I think this 'highlighted' style is used also elsewhere. It should be fixed wherever it is used.

Screenshot: hint: the fourth suggestion is the highlighted one.

<img width="518" alt="Screenshot 2023-06-13 at 15 23 57" src="https://github.com/WordPress/gutenberg/assets/1682452/b0cdc41c-eda2-4f7c-a4c5-821a02fc7a5a">

### Step-by-step reproduction instructions

- Go to teh Site editor.
- Open the Command centre.
- Enter a search term e.g.: 'template'.
- Wait for the list of suggestions to be displayed.
- Use the Down and Up arrow keys to navigate the suggestions.
- Also try hovering the suggestions.
- Observe the highlighted style only uses a color change.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by jameskoster** [comment]
Previous mockups for suggestions have included a classifying label, e.g. "Command", "Shortcut", etc:

<img width="387" alt="Screenshot 2023-06-19 at 13 34 48" src="https://github.com/WordPress/gutenberg/assets/846565/79d8d2ce-bebd-4d4b-adf4-29cf829c1272">

If such as label was iconised and exposed on hover, would that qualify as a shape change?

<img width="541" alt="Screenshot 2023-06-19 at 13 40 31" src="https://github.com/WordPress/gutenberg/assets/846565/84bf9b41-833c-462a-a95d-f9720595c0d7">

**Comment by jasmussen** [comment]
I suggested a "Run command" text a la GitHub on [the PR](https://github.com/WordPress/gutenberg/pull/51626#issuecomment-1597143678),  but I like the arrow better. I wonder if it should be a chevron? It could also be a little [Run] chip design. But yes, it qualifies as a shape change.

**Comment by jameskoster** [comment]
At the moment we have the following 'types' of actions:

* Shortcut – E.g. go directly to the Styles panel
* Command – E.g. Delete a template, Add a page

It _might_ get a little fuzzy if we have different icons for each? So I would be tempted to stick to a single icon to begin with, if we go down that route and revisit if/when more action types are added. I chose the Arrow instead of the Chevron since we use chevrons to indicate drilldowns elsewhere, but maybe that doesn't matter? Or maybe there's another icon to try.

Text labels would probably be clearer: [Run] for commands, [Jump to] (or similar) for shortcuts.

<img width="545" alt="Screenshot 2023-06-20 at 10 46 24" src="https://github.com/WordPress/gutenberg/assets/846565/348a857a-6860-4b71-8ba5-b12d4e5957ab">

That would enable us to remove the 'Open...' prefixes too which are a little ambiguous imo. For that reason I'm starting to lean more towards the text labels 😅

**Comment by jasmussen** [comment]
Single icon sounds like a good start and is easy to implement, and doesn't preclude revisits in the future.

**Comment by jameskoster** [comment]
Yep, that's a good point. Alternatives to the arrow:

### Return
<img width="554" alt="Screenshot 2023-06-20 at 11 23 21" src="https://github.com/WordPress/gutenberg/assets/846565/8eee0a6f-4b30-49ae-95d7-3297a6675e26">

Could be a hint for keyboard interaction?

### Chevron

<img width="554" alt="Screenshot 2023-06-20 at 11 25 01" src="https://github.com/WordPress/gutenberg/assets/846565/628363c5-2d63-43e5-a3b5-df282363af02">

I worry this suggests a drilldown which isn't what happens. Not a stong feeling, but maybe we should reserve this in case drilldowns are added to the command palette down the road?

**Comment by jasmussen** [comment]
Good point on the chevron. I don't personally share the full concern, though I would think the regular arrow works better than the return.

**Comment by afercia** [comment]
While the arrow could work, I'd rather suggest to look at the big picture and try to standardize this 'highlighted option' style across the UI. As far as I can tell there are at least two components that show a list of options:
- Combobox (search + suggested options)
- CustomSelectControl

where the highlighted style is largely inconsistent. A few examples:

Slash inserter:

<img width="431" alt="slash inserter" src="https://github.com/WordPress/gutenberg/assets/1682452/15e50f35-5243-41a0-8f1e-1a0a44cefa73">

Link suggestions:

<img width="470" alt="link suggestions" src="https://github.com/WordPress/gutenberg/assets/1682452/f0445fc8-9f17-4499-83e2-95e57ca7420f">

Tag suggestions:

<img width="297" alt="tags suggestions" src="https://github.com/WordPress/gutenberg/assets/1682452/d74a6d1e-274e-49b1-94a3-41de5c44ea41">

Position settings:

<img width="321" alt="position" src="https://github.com/WordPress/gutenberg/assets/1682452/4a3020f9-b4ff-469b-84d4-6ff6173433fc">

Comment date format:

<img width="307" alt="comment date format" src="https://github.com/WordPress/gutenberg/assets/1682452/8590c23b-2ed7-47a4-807c-fe04bed8c0a4">

Font appearance:

<img width="304" alt="font appearance" src="https://github.com/WordPress/gutenberg/assets/1682452/ff4d783f-b5f3-4cea-afac-94fe79f6753f">

**Comment by jameskoster** [comment]
Since (as pointed out in the OP) this isn't a true focus, wouldn't it be a little contradictory to use the outline treatment?

Perhaps it would help to better define 'shape change' for this context? IE why is adding a background color not a shape change? Isn't it the addition of a shape where none was previously visible? Would it be different if the background had stronger contrast, IE:

<img width="517" alt="Screenshot 2023-06-21 at 11 05 24" src="https://github.com/WordPress/gutenberg/assets/846565/c61fc446-e41d-4e83-aca1-67f4aeb89a4f">

To be clear I'm not proposing we do that, just trying to better understand the problem. Thanks.

**Comment by richtabor** [comment]
> I worry this suggests a drilldown which isn't what happens. 

We may have drilldown commands one day too.

**Comment by afercia** [comment]
After two months this issue has been created this still needs a design decision. In the meantime, the command center has been realeased in WordPress 6.3. It's unfortunate that a such prominent feature gets released with an insufficient accessibility (and usability) level. Honestly, I'm not sure this way of releasing software is ideal for a project that aims to match a high quality level.

That said, this only needs a design decision. The implementation part is just CSS. The challenge here is mostly about making all the autocomplete comboboxes look the same across the editor. As of now, there's still large inconsistency across all the autocompleters and I'm no sure this is ideal for users and for the reputation of WordPress.

Couple screenshots from current trunk:

Command center: the highlighted suggestion is barely visible:

<img width="462" alt="Screenshot 2023-08-22 at 09 44 14" src="https://github.com/WordPress/gutenberg/assets/1682452/d8fd80ec-540a-4bb8-bb51-9113ccd2829a">

Tags: totally different styling and also functionality, as the search term is underlined within the suggestions:

<img width="330" alt="Screenshot 2023-08-22 at 09 46 40" src="https://github.com/WordPress/gutenberg/assets/1682452/6c4dda98-139a-4bbc-aa58-e49f8c1d9831">

**Comment by joedolson** [comment]
We shouldn't need to wait on a global design decision in order to fix an extremely obvious accessibility problem. We can resolve the accessibility problem with a very simple design change to improve contrast; all of the other larger design issues can be resolved after fixing the accessibility issue.

We need to keep in mind that fixing an accessibility issue has a direct and immediate impact on users who depend on it; it resolves a major user experience problem for the people it impacts. The other design issues need to be done with accessibility in mind, but should not block us from fixing an immediate problem.

**Comment by jeryj** [comment]
To close this issue, we only need a decision on the colors to use to highlight the suggestion. Those colors need to have sufficient contrast. The other design decisions on icons and consistency across the editor can be a follow-up.

**Comment by richtabor** [comment]
> Since (as pointed out in the OP) this isn't a true focus, wouldn't it be a little contradictory to use the outline treatment?

This unanswered question is what's holding up. 

> We shouldn't need to wait on a global design decision in order to fix an extremely obvious accessibility problem. 

We should always push for consistency, rather than ad-hoc fixes (which results in varying experience like expressed in [this comment above](https://github.com/WordPress/gutenberg/issues/51582#issuecomment-1600355521)  — but I do agree we need to resolve this. 


That saying, here's what we have today:

<img width="673" alt="current" src="https://github.com/WordPress/gutenberg/assets/1813435/36f869a4-024a-4ff9-b5fc-ee0767e7038d">

And below are the options:

<img width="673" alt="2" src="https://github.com/WordPress/gutenberg/assets/1813435/a3fb1b2c-035d-42f8-a400-c97f393d6206">

<img width="673" alt="1" src="https://github.com/WordPress/gutenberg/assets/1813435/8940de7f-7fb0-4b1f-9abc-60aa0e596731">

<img width="673" alt="4" src="https://github.com/WordPress/gutenberg/assets/1813435/6dcd7d44-5356-49f4-be30-583831cdc9ee">

<img width="673" alt="5" src="https://github.com/WordPress/gutenberg/assets/1813435/9979e7e0-093d-429f-b33a-9bc204edca06">

<img width="668" alt="CleanShot 2023-09-08 at 15 24 51" src="https://github.com/WordPress/gutenberg/assets/1813435/39f1d954-7175-43e0-9228-8f98c6c8e9d4">


### Request
The options above are based on other variants of existing flows within the editor. 

Do you mind detailing **what won't work** (and why), as well as **what will work** (and why). That would help clear the air and move this forward (as well as consolidating the other suggestion + selection flows). 

I would like to come to a solution that we can apply consistently everywhere, so in the future when something like this comes up there is a clear standard. cc @WordPress/gutenberg-design

**Comment by joedolson** [comment]
Any of the combinations that meet basic accessibility standards for color contrast will work. Beyond that, I have no concern about which design is used. Not knowing which colors are represented above, I can't say with certainty which of those meet standards.

**Comment by jeryj** [comment]
> Since (as pointed out in the OP) this isn't a true focus, wouldn't it be a little contradictory to use the outline treatment?
> This unanswered question is what's holding up.

I don't think there's a standard around this, and it's more about consistency and color contrast than which specific style is chosen. That said, since the outline focus is used on the inputs and sometimes wrapper depending on context, that having a solid color background to show selection will be the best choice. Whichever solid color background that meets the color contrast guidelines (4.5:1 for AA standard) sounds good to me. 

The purple one is appealing, as it has strong contrast with both the text inside the color and against the white background. It also feels a bit more modern and bold to me. The light gray background makes it harder to see which one is selected since there's not much contrast between the light gray and white background.

**Comment by richtabor** [comment]
@joedolson I'm trying to work _with_ you and move this forward. I personally have concern for accessibility _and_ design—we do need both. 

> Not knowing which colors are represented above, I can't say with certainty which of those meet standards.

All options are placed on a `#FFFFFF` canvas: 

Option one: 
- Background `#FFFFFF`
- Foreground `#3858E9` (based on current admin color scheme)

Option two:
- Background `#3858E9` (based on current admin color scheme)
- Foreground `#FFFFFF`

Option three: 
- Background `#F0F0F0`
- Foreground `#1E1E1E`

Option four: 
- Background `#E0E0E0`
- Foreground `#1E1E1E`

Option five:
- Background `#FFFFFF`
- Foreground `#3858E9`  (based on current admin color scheme)
- Outline `#3858E9`  (based on current admin color scheme)

**Comment by jeryj** [comment]
Did some Googling to confirm. I could be wrong, but I believe that to meet AA guidelines the color combinations need to pass two checks:
- The color of the text on the selected item background needs to be 4.5:1 or higher
- The background color of the selected item and the background color of the command palette need to be 3:1 or higher.

Out of those options, only option 2 meets those guidelines. The others don't meet the 3:1 contrast guideline between the selection background and command palette background.

**Comment by joedolson** [comment]
@richtabor  I don't have any personal preference about the design choices. I favor the inverted contrast  (2) and the outline option (5), because they offer a higher degree of visual affordance of the selection. I apologize for being curt in my previous response.

@jeryj's description is accurate, as far as contrast is concerned. When color is the only distinguishing factor, both contrast requirements need to be met: the first describes the ability to see the text against it's background (readability) and the second describes the ability to distinguish which item is selected. 

However, because there's an icon present that *also* indicates which item is selected, the second color distinction isn't necessary - that color contrast distinction is only required if color is the only differentiation between selected and unselected items. What Jerry is referring to is success criterion [1.4.11: Non-text contrast.](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html)

If the selection icon was also below contrast requirements, that would be a problem.

**Comment by jeryj** [comment]
Ah, so if there's an icon to demonstrate _selection_ then we don't need a different background color to show selection. I think in this case though, the "Open in a new tab" icon isn't to show selection but the icon that is associated with that item. Here's an example of the command palette search results with more item types, each with their own icon. I'm guessing we wouldn't want to add another icon to the mix to show selection.

<img width="424" alt="Temp search on command palette with results of varying icons" src="https://github.com/WordPress/gutenberg/assets/967608/b2229bc9-3433-496e-a7e9-deb4e0fdb12f">

Would an outline to show selection meet AA guidelines here?
![Blueberry outline, no text color change to show selection](https://github.com/WordPress/gutenberg/assets/967608/fc478481-42c4-4dca-ada6-ef6c8faae0c1)

**Comment by richtabor** [comment]
>  I think in this case though, the "Open in a new tab" icon isn't to show selection but the icon that is associated with that item. 

Correct. Icons are assigned when a command is registered. 


>  I'm guessing we wouldn't want to add another icon to the mix to show selection.

Yea, I'd rather lean on color to indicate selection rather than more iconography.

**Comment by joedolson** [comment]
In that case, yes - only option 2 achieves the needed goal. Based on the examples, it seemed like the icon was being used to indicate which item was active, but if that's not the case, we'll need to make sure all the color contrast baselines are met.
